gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,175 @@
+        """
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        if num_images > MAX_NUM_IMAGES:
+            raise RuntimeError("Cannot construct similarity matrix of this size.")
+
+        sim = torch.zeros((num_images, num_images))
+        num_blocks = math.ceil(num_images / self._blocksize)
+
+        for block_i in range(num_blocks):
+            for block_j in range(num_blocks):
+                # only compute the upper triangular portion of the similarity matrix.
+                if block_i > block_j:
+                    continue
+                logger.info("Computing matching block (%d/%d,%d/%d)", block_i, num_blocks - 1, block_j, num_blocks - 1)
+                block_i_query_descs = []
+                block_j_query_descs = []
+
+                i_start = block_i * self._blocksize
+                i_end = (block_i + 1) * self._blocksize
+                i_end = min(i_end, num_images)
+
+                j_start = block_j * self._blocksize
+                j_end = (block_j + 1) * self._blocksize
+                j_end = min(j_end, num_images)
+
+                block_i_idxs = np.arange(i_start, i_end)
+                block_j_idxs = np.arange(j_start, j_end)
+
+                for i in block_i_idxs:
+                    image = loader.get_image(i)
+                    block_i_query_descs.append(self._global_descriptor_model.describe(image))
+
+                for j in block_j_idxs:
+                    image = loader.get_image(j)
+                    block_j_query_descs.append(self._global_descriptor_model.describe(image))
+
+                # Form (K,D) for K images.
+                block_i_query_descs = torch.from_numpy(np.array(block_i_query_descs))
+                block_j_query_descs = torch.from_numpy(np.array(block_j_query_descs))
+
+                # Einsum equivalent to (img_descs @ img_descs.T)
+                sim_block = torch.einsum("id,jd->ij", block_i_query_descs.to(device), block_j_query_descs.to(device))
+                sim[i_start:i_end, j_start:j_end] = sim_block
+
+        return sim
+
+
+def pairs_from_score_matrix(
+    scores: torch.Tensor, invalid: np.array, num_select: int, min_score: Optional[float] = None
+) -> List[Tuple[int, int]]:
+    """Identify image pairs from a score matrix.
+
+    Args:
+        scores: (K1,K2) for matching K1 images against K2 images.
+        invalid: (K1,K2) boolean array indicating invalid match pairs (e.g. self-matches).
+        num_select: number of matches to select, per query.
+        min_score: minimum allowed similarity score.
+
+    Returns:
+        pairs: tuples representing pairs (i1,i2) of images.
+    """
+    N, _ = scores.shape
+    # if there are only N images to choose from, selecting more than N is not allowed
+    num_select = min(num_select, N)
+
+    assert scores.shape == invalid.shape
+    invalid = torch.from_numpy(invalid).to(scores.device)
+    if min_score is not None:
```

Comment:
Should we at least require that the scores are greater than some epsilon? It might not be a good idea to return pairs whose descriptors are almost perpendicular.

Reply:
We probably should. I plotted what theshold would be good for skydio-250, and saw 0.5 as a good threshold. That threshold doesn't work for Door-12, though. So a bit hard to tune.
![Screen Shot 2022-02-10 at 10 49 30 AM](https://user-images.githubusercontent.com/16724970/153717990-88b0e246-6216-4b3a-a090-b987f7a63c69.png)
![Screen Shot 2022-02-10 at 10 49 17 AM](https://user-images.githubusercontent.com/16724970/153717992-0090de19-f45f-44cb-8c3a-944e11cbb3a2.png)

Reply:
These plots are great; if they're not already generated automatically in the codebase we should add them.

Is it Skydio-250 top and door-12 bottom?

Reply:
Thanks These are actually both from Skydio-250 (U error plot, and then R error plot).

Sure, I can add that -- it's only about 10 lines of code to call at the end.

Reply:
I added this code now to `scene_optimizer.py` as a free function, called at the end of execution.